src/compile_cache.cc

```diff
@@ -223,13 +224,67 @@ void CompileCacheHandler::ReadCacheFile(CompileCacheEntry* entry) {
   Debug(" success, size=%d\n", total_read);
 }
 
+static std::string NormalisePath(std::string_view path) {
+  std::string normalised_path(path);
+  constexpr std::string_view file_scheme = "file://";
+  if (normalised_path.rfind(file_scheme, 0) == 0) {
+    normalised_path.erase(0, file_scheme.size());
+  }
+  normalised_path = NormalizeString(normalised_path, false, "/");
+  return normalised_path;
+}
+
+inline bool HasPathPrefix(std::string_view path, std::string_view prefix) {
+  if (path.size() < prefix.size()) {
+    return false;
+  }
+#ifdef _WIN32
+  if (_strnicmp(path.data(), prefix.data(), prefix.size()) != 0) {
+    return false;
+  }
+#else
+  if (path.compare(0, prefix.size(), prefix) != 0) {
+    return false;
+  }
+#endif
+  if (path.size() == prefix.size()) {
+    return true;
+  }
+  return path[prefix.size()] == '/';
+}
+
 CompileCacheEntry* CompileCacheHandler::GetOrInsert(Local<String> code,
                                                     Local<String> filename,
                                                     CachedCodeType type) {
   DCHECK(!compile_cache_dir_.empty());
 
   Utf8Value filename_utf8(isolate_, filename);
-  uint32_t key = GetCacheKey(filename_utf8.ToStringView(), type);
+  std::string normalised_file_path =
+      NormalisePath(filename_utf8.ToStringView());
+  std::string key_path = normalised_file_path;
+  std::string current_path;
+
+  if (use_relative_) {
+    char cwd[PATH_MAX_BYTES];
+    size_t size = PATH_MAX_BYTES;
+    if (uv_cwd(cwd, &size) == 0 && size > 0) {
```

Comment:
This can be very expensive.

I wonder if we can just use some heuristics to check if the filename _looks like_ an absolute path or a file URL. Ordinary modules generally should have absolute filenames/URLs passed here at this point. That should just trim down the computation to a few string searches like `file://` or starting with `/` or substring from 1-2 is `:\` or `:/`

Reply:
Also, if the relative path computation is too laborious to reimplement, I think it's fine to use `std::filesystem::path::lexically_relative`, just make sure that it's always correctly transcoded into UTF16 on Windows, and converted back to UTF8 when being used as a cache key. I _think_ Windows is the only one that has this caveat.